File: js/array/longest-consecutive-sequence.js
/**
 * https://leetcode.com/problems/longest-consecutive-sequence
 * 
 * @param {number[]} nums
 * @return {number}
 */
const longestConsecutive = (nums) =>  {

    if (!nums.length) return 0;

    const uniqueNums = new Set(nums);
    let count = 1;
    let maxSeq = 1;

    for (value of uniqueNums) {
        let curr = value
        if (uniqueNums.has(curr - 1)) {
            continue;
        } else {
            while (uniqueNums.has(curr + 1)) {
                count++;
                curr++;
            };

            maxSeq = Math.max(maxSeq, count);
            count = 1;
        }
    };
    return maxSeq;
};

File: python/array/longest-consecutive-sequence.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:

        if len(nums) == 1:
             return 1

        num_exists = set(nums)

        maxCount = 0
        for i in num_exists:
            count = 1
            if(i - 1) not in num_exists:
                count = 0
                setVal = i
                while(setVal) in num_exists:
                    setVal += 1
                    count += 1
                maxCount = max(count, maxCount)


        return maxCount

            

# https://leetcode.com/problems/longest-consecutive-sequence
# Hard moved to medium, solution asks you not to sort (as it wants o(n), not o(n log n))
# To solve, cast array to set for lookup operations, then for each number, if it doesn't
# have a number before it, it is the start of a sequence, so forward search set.